### ISSUE TITLE
Data lost when upgrading from 4.0.0 to 5.0.1

### ISSUE THREAD

**Issue by andre-arsenault** [felangel/hydrated_bloc#67]
**Describe the bug**
After updating an app's dependencies to use hydrated_bloc 5.0.1, all saved data is lost from previous runs of the app. Verified on Android and iOS.

**To Reproduce**
1. Run your app with hydrated_bloc 4.0.0
2. Make some state changes which are saved via hydrated_bloc
3. Re-run the app to see that the state changes were saved from the previous run
4. Update the app to use hydrated_bloc 5.0.1
5. Run the app and see that none of the previous state is retained

**Expected behavior**
Updating from 4.0.0 to 5.0.1 should be a non-destructive operation for users.

**Logs**

```
andre@Andres-MacBook-Pro gaming_news % flutter doctor -v
[✓] Flutter (Channel beta, 1.19.0-4.3.pre, on Mac OS X 10.15.5 19F101, locale en-US)
    • Flutter version 1.19.0-4.3.pre at /Users/andre/Development/flutter
    • Framework revision 8fe7655ed2 (5 days ago), 2020-07-01 14:31:18 -0700
    • Engine revision 9a28c3bcf4
    • Dart version 2.9.0 (build 2.9.0-14.1.beta)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /Users/andre/Library/Android/sdk
    • Platform android-29, build-tools 28.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.5)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.5, Build version 11E608c
    • CocoaPods version 1.9.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 3.6)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 44.0.2
    • Dart plugin version 192.7761
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)

[✓] VS Code (version 1.46.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.12.1

• No issues found!
```

**Additional context**
After some time investigating the issue, it is caused by the underlying migration to the hydrated_cubit library. That migration in itself isn't the problem, but in the commit to switch to hydrated_cubit, the backwards-compatible migration code was deleted.

The file storage used for hydration in 4.0.0 was `<directory>/.hydrated_bloc.json` but was moved to `<directory>/hydrated_box.hive[c]` in 5.0.1. Commit 7cae45f8d83037250541865e69b8ebbce83eb2d0 had migration code for loading `.hydrated_bloc.json` if it existed. 

See:

- Migration code added: https://github.com/felangel/hydrated_bloc/commit/7cae45f8d83037250541865e69b8ebbce83eb2d0#diff-03e6729a69de9f7669463016383f33f8
- Migration code deleted: https://github.com/felangel/hydrated_bloc/commit/07261573db71cfdb3aa054099c167be9ecf5f109#diff-03e6729a69de9f7669463016383f33f8

**Workaround**
For other users experiencing this issue and needing a timely resolution, you can use you own custom hydrated storage as described in the README.

The following workaround works for my use case with 5.0.1. It is based on `HydratedStorage` from 4.0.0 and does not use hive for storage. There are key changes to the `read` and `write` methods to support differences introduced in 5.0.0.

``` dart
class HydratedBlocStorageWorkaround implements HydratedStorage {
  HydratedBlocStorageWorkaround._(this._storage, this._file);

  static final _lock = Lock();
  final Map<String, dynamic> _storage;
  final File _file;

  static Future<HydratedBlocStorageWorkaround> getInstance({
    Directory storageDirectory,
  }) {
    return _lock.synchronized(() async {
      final directory = storageDirectory ?? await getTemporaryDirectory();
      final file = File('${directory.path}/.hydrated_bloc.json');
      var storage = <String, dynamic>{};

      if (await file.exists()) {
        try {
          storage = json.decode(await file.readAsString()) as Map<String, dynamic>;
        } on dynamic catch (_) {
          await file.delete();
        }
      }

      return HydratedBlocStorageWorkaround._(storage, file);
    });
  }

  @override
  dynamic read(String key) {
    final String jsonString = _storage[key];
    return jsonString?.isNotEmpty == true
        ? Map<String, dynamic>.from(json.decode(jsonString))
        : null;
  }

  @override
  Future<void> write(String key, dynamic value) {
    return _lock.synchronized(() {
      _storage[key] = json.encode(value);
      return _file.writeAsString(json.encode(_storage));
    });
  }

  @override
  Future<void> delete(String key) {
    return _lock.synchronized(() {
      _storage[key] = null;
      return _file.writeAsString(json.encode(_storage));
    });
  }

  @override
  Future<void> clear() {
    return _lock.synchronized(
      () async {
        _storage.clear();
        if (await _file.exists()) {
          await _file.delete();
        }
      },
    );
  }
}
```

**Comment by felangel** [comment]
Hi @andre-arsenault 👋
Thanks for opening an issue and sorry for any inconvenience!

If you downgrade to 4.1.1 and then upgrade to 5.0.0 then no data will be lost (see https://github.com/felangel/hydrated_bloc/issues/66).

Let me know if that helps and again sorry for any inconvenience.

**Comment by andre-arsenault** [comment]
Hi @felangel , thanks for the response!

That involves downgrading all other bloc-related packages as well, correct?

```
[gaming_news] flutter pub get
Running "flutter pub get" in gaming_news...                     
Because no versions of bloc_test match >6.0.0 <7.0.0 and bloc_test 6.0.0 depends on bloc ^5.0.0, bloc_test ^6.0.0 requires bloc ^5.0.0.

And because hydrated_bloc >=4.0.0 <5.0.0-dev.1 depends on bloc ^4.0.0, bloc_test ^6.0.0 is incompatible with hydrated_bloc >=4.0.0 <5.0.0-dev.1.

So, because gamescope depends on both hydrated_bloc ^4.1.1 and bloc_test ^6.0.0, version solving failed.
pub get failed (1; So, because gamescope depends on both hydrated_bloc ^4.1.1 and bloc_test ^6.0.0, version solving failed.)
exit code 1
```

It's not an ideal solution regardless of the dependency entanglements, as we would need to:

- deploy a version of our app using 4.1.1
- wait for as many users as possible to upgrade
- deploy a new version using 5.0.1
- any users who didn't upgrade to the intermediate 4.1.1 version will have their app state erased

That's a lot to ask when a few lines of migration code could maintain backward compatibility for 100% of users.

Would you consider changing your stance from #66? I could ship my workaround but I presume there are good reasons to use hive for persistence going forward.

**Comment by felangel** [comment]
@andre-arsenault yup that’s totally understandable and I can reintroduce the migration code in 5.0.2 of hydrated_bloc 👍

Should have the fix in shortly and again sorry for the trouble.

**Comment by andre-arsenault** [comment]
That's great @felangel, thanks very much!

**Comment by felangel** [comment]
PR is out and ready for review at https://github.com/felangel/cubit/pull/61

**Comment by felangel** [comment]
@andre-arsenault can you confirm that the issue is resolved by referencing the 5.0.2 release candidate via git url?

```yaml
hydrated_bloc:
  git:
    url: https://github.com/felangel/hydrated_bloc
    ref: 118937c68fd0106106ba177cab84b4a7f89fc469
```

Thanks!

**Comment by andre-arsenault** [comment]
@felangel confirmed! That version is able to load 4.0.0 data and saves it in 5.0.1 format. That's a big help, thank you! 🎉

**Comment by felangel** [comment]
@andre-arsenault the fix has been published in v5.0.2 🎉

**Comment by andre-arsenault** [comment]
That's great, thanks for the swift reaction! Very much appreciated!
